### ISSUE TITLE
support for vulkan video decode/encode apis

### ISSUE THREAD

**Issue by maddanio** [KhronosGroup/MoltenVK#1337]
Hello

will MoltenVK implement the new vulkan video encode/decode apis?

https://www.khronos.org/blog/an-introduction-to-vulkan-video

best

**Comment by cdavis5e** [comment]
While I'm pretty sure Metal already has support for hardware-accelerated video encoding/decoding, until Apple exposes it publicly, there isn't much I think we can do until then.

**Comment by cdavis5e** [comment]
Actually... we might be able to use the VideoToolbox framework to support this.

**Comment by maddanio** [comment]
yes, that was the idea. we are already doing vulkan<->vt exchange with metal surfaces. if it helps I can isolate the neccessary steps that are proven to work for us.

**Comment by maddanio** [comment]
it's not particularly urgent for us, especially as long as support for this is still sparse elsewhere (intel, amd), but it came up and I felt it should be a relatively mechanical task to implement and nice to support the initiative by khronos

**Comment by mbechard** [comment]
I'll mention that we'd find this super useful. Being able to use hardware encode/decode via a single API across multiple platforms will be really helpful.

**Comment by oscarbg** [comment]
+1

**Comment by ThomasHagenJohansen** [comment]
+1

**Comment by maddanio** [comment]
so, let's see, been a while :). so for the decode side

1. create a CVMetalTextureCache for the device
2. given a CVImageBufferRef (which you get from core video) create one metal texture per plane using CVMetalTextureCacheCreateTextureFromImage
3. create a VKImage for each plane that is compatible with the right properties
4. use vkSetMTLTextureMVK to set the image to the texture obtained via CVMetalTextureGetTexture

for the encoding side its basically the same, only then you render to the obtained texture on the vulkan side and use the pixel puffer on the core video side

i hope this helps. I can also try to isolate some code, but that may take longer, as its tangled up in our codebase a bit

**Comment by maddanio** [comment]
ah and currenlty this means you have to link directly to moltenvk and not use any layers, unless they made it possible now to "see through" layers as I think was requested a while ago

**Comment by maddanio** [comment]
Any updates?
